[ModuleTrace] Handle cycle detection edge case in module trace emission.

varungandhi-apple opened the pull request:
Fixes rdar://66512316.

varungandhi-apple committed:
[NFC] [ModuleTrace] Add negative test for cycles in trace emission.

Context: rdar://66512316.

varungandhi-apple committed:
[ModuleTrace] Handle cycle detection edge case in module trace emission.

Fixes rdar://66512316.

varungandhi-apple commented:
@swift-ci please smoke test

brentdax commented on lib/FrontendTool/FrontendTool.cpp:
```diff
@@ -453,9 +482,44 @@ void ABIDependencyEvaluator::reexposeImportedABI(
+      llvm::any_of(importList, [&](ModuleDecl *importedModule) -> bool {
+        return isClangOverlayOf(swiftModule, importedModule);
+      });
+  isOverlayCache[swiftModule] = isOverlay;
```
How expensive is this? Naïvely, it looks like we're caching an O(n) computation; is it hot enough to be worth the complexity of the cache?

varungandhi-apple replied:
The complexity is incurred on a per-import basis. This means that if I have `N_o` overlays, each such overlay has `N_i` import statements and each overlay is imported by `N_c` Clang modules, then we will have a `N_c * N_i * N_o` computation instead of a `N_i * N_o` computation.

varungandhi-apple replied:
I measured this with an example module which imports many modules (and does just that); the cache saves about 0.2 ms ~ 0.4 ms per trace. Which doesn't seem like much at all, so I'm going to get rid of it.

brentdax commented on lib/FrontendTool/FrontendTool.cpp:
```diff
@@ -453,9 +482,44 @@ void ABIDependencyEvaluator::reexposeImportedABI(
+}
+
+bool ABIDependencyEvaluator::isFakeCycleThroughOverlay(
+    ModuleDecl **sandwichModuleIter) {
```
I might defensively assert that `sandwichModuleIter` is an element in `searchStack`, since that's what guarantees that all the dereferencing and pointer arithmetic is valid:

```cpp
  assert(sandwichModuleIter >= searchStack.begin() && sandwichModuleIter < searchStack.end()
         && "sandwichModuleIter points to an element in searchStack");
```

But that might just be me being paranoid.

varungandhi-apple replied:
Paranoia is good. Little bit of arithmetic won't hurt anyone.

varungandhi-apple commented:
@artemcm you might be interested in looking at the test case here. This is kinda' what I was talking about when I brought up that explicit module builds would need to handle dependencies through Clang modules and generated headers.

beccadax reviewed (changes requested):
I have some minor comments, but the core of this looks sound.

varungandhi-apple commented on lib/FrontendTool/FrontendTool.cpp:
```diff
@@ -342,6 +345,32 @@ class ABIDependencyEvaluator {
+  /// \code
+  ///     Overlay (Swift) -> sandwichedModule -> Underlying (Clang)
+  /// \endcode
+  /// The stack would look like:
```
I shuffled the comment a bit; forgot to delete this. Oops.

varungandhi-apple commented:
@swift-ci please smoke test

varungandhi-apple committed:
[NFC] Remove caching for "isOverlay" computation when emitting a trace.

Based on measurements, it seemed to save about 0.2 ms - 0.4 ms for a module
with a large number of imports from the SDK. That's insignificant compared to
the compile time for a typical module, which usually takes somewhere between
a few seconds and a few minutes.

varungandhi-apple commented:
@swift-ci please smoke test

hamishknight commented:
The module trace emission logic seems to be getting fairly long – would it be worth splitting it into its own file at some point?

varungandhi-apple commented:
Sounds like a reasonable change for a follow-up. I've made note of it.
